client/protocols/guidance.py

```diff
@@ -0,0 +1,178 @@
+            # if student is in control group, just print the default message
+            print ("-- Not quite. Try again! --")
+            return (self.misU_count, self.tg_id,"")
+
+        print("-- Helpful Hint --\n")
+
+        printed_out_msgs = ""
+        for mid in self.msg_id_set:
+            msg = self.guidance_json['dictId2Msg'][str(mid)]
+            printed_out_msgs = printed_out_msgs + msg
+            print(msg)
+
+        print("\n-- Helpful Hint --")
+
+        return (self.misU_count, self.tg_id,printed_out_msgs)
+
+    # Looks at the locally saved file for number of misU and returns the current count
+    def update_misUcounts(self, hashkey, misU, wrongAnswer, shorten_unique_id):
+        #Creates a new folder inside tests that stores the number of misU per assignment
+        count_file_path = "tests/misUcount.json"
+        if os.path.isfile(self.working_dir + count_file_path):
+            with open(self.working_dir + count_file_path,'r') as f:
+                jsonDic = json.load(f)
+                answerDict = jsonDic["answerDict"]
+                countData = jsonDic["countData"] 
+        else:
+            countData = {}
+            answerDict = {}
+        newjsonDic = {}
+
+        # Checks to see if the question is already stored and whether the student's answer is there
+        # It does not update misU count if same answer was seen before
+        if shorten_unique_id in answerDict and wrongAnswer in answerDict[shorten_unique_id]:
+            return countData
+        if not shorten_unique_id in answerDict:
+            answerDict[shorten_unique_id] = []
+
+        # Updates misU count
+        for x in misU: 
+            if x in countData:
+                countData[x] += 1
+            else:
+                countData[x] = 1
+
+        # Stores the updated count back into the same file and overrides it
+        answerDict[shorten_unique_id].append(wrongAnswer)
+        newjsonDic["countData"] = countData
+        newjsonDic["answerDict"] = answerDict
+        with open(self.working_dir + count_file_path,"w") as f:
+            json.dump(newjsonDic,f)
+        # Returns a dictionary containing how many times each misU has been seen.
+        return countData
+
+    def get_tg(self,access_token,guidance_flag):
+        if guidance_flag:
+            with open(self.working_dir + LOCAL_TG,"w") as f:
+                f.write(str(guidance_flag_dict[0]))
+            self.tg_id = guidance_flag_dict[0]
+            return
+        # Checks to see the student currently has a treatment group number. If not, calls helper function in auth.py
+        if not os.path.isfile(self.working_dir + LOCAL_TG):
+            cur_email = auth.get_student_email(access_token)
+            if not cur_email:
+                print ("--Not quite. Try again! --")
+                return ({}, -1)
+            try:
+                data = json.loads(urlopen(TGSERVER + cur_email + "/unlock_tg",timeout =1).read().decode("utf-8"))
```

Comment:
Also I thought the final version would also include information on the lab, in case we do treatment groups based on the person and the lab. Technically we won't do this for the pilot, but if it's a simple code change, might as well.